### PR TITLE
CB2-7573 Handle changes to test result test types and defects

### PR DESCRIPTION
**Pull request by Ian-Hodges**
## Description

**CB2-7573:** When a test is modified and reprocessed into NOP, it can result in the old and new result being attached to the record. This PR adds code to delete results if they already exist prior to processing the updated record.
**CB2-7685:** The test_result table has a trigger that populates the testtype_version table. It has been found that it can affect the identity returned to the update store lambda. This PR has changes to address this.

Related issue: [CB2-7573](https://dvsa.atlassian.net/browse/CB2-7573)
Related issue: [CB2-7685](https://dvsa.atlassian.net/browse/CB2-7685)

The main changes are in the update-store lambda. Details of the evidence of completion can be seen in [cvs-tsk-update-store/pull/64](https://github.com/dvsa/cvs-tsk-update-store/pull/64) 

### Clean pipeline build
image

### CODE REVIEW

File: sql/00000_base_database.sql [hunk 1]
```diff
@@ -556,6 +557,7 @@ CREATE TABLE IF NOT EXISTS test_result
 
     INDEX `idx_vehicle_id` (`vehicle_id` ASC),
     INDEX `idx_test_number_id` (`testNumber` ASC),
+    INDEX `idx_test_result_id` (`testResultId` ASC),
```
Comment (ChrisPeska-BJSS) [inline sql/00000_base_database.sql:560]: Would the additional index be to support `enquiry` service GET requests?

File: sql/00000_base_database.sql [hunk 2]
```diff
@@ -693,13 +696,15 @@ CREATE TABLE IF NOT EXISTS `testtype_version`
     `lastSeatbeltInstallationCheckDate` DATE,
     `seatbeltInstallationCheckDate`     TINYINT(1),
     `testResult`                        VARCHAR(9),
-    `reasonForAbandoning`               VARCHAR(45),
+    `reasonForAbandoning`               VARCHAR(420),
```
Comment (ChrisPeska-BJSS) [inline sql/00000_base_database.sql:699]: More of the mental note. That would be a separate change I imagine, I think we saw an outlier in Prod data with length exceeding 1000 characters if I recall correctly.

File: sql/00003_test_result_audit_test_type_version_trigger.sql [hunk 3]
```diff
@@ -0,0 +1,35 @@
+--liquibase formatted sql
+--changeset liquibase:create -multiple-tables:1 splitStatements:true endDelimiter:// context:dev
+CREATE TRIGGER check_tt_version_delete AFTER DELETE ON `test_result`
```
Comment (ChrisPeska-BJSS) [inline sql/00003_test_result_audit_test_type_version_trigger.sql:3]: Once coupled with the [change in update-store ](https://github.com/dvsa/cvs-tsk-update-store/compare/develop...feature/CB2-7573#diff-2b9ea8b3f1402f58aaf24d001c13e90268b474e9234742b05eb69e489bbecbf3R140)the trigger would register any update to Test Result table?

**Review by ChrisPeska-BJSS — APPROVED** [review]
I did review it in line with Pull Request https://github.com/dvsa/cvs-tsk-update-store/pull/64.

I have only one little concern around ever growing `testtype_version` table, as there is no retention on this table. Unless I may not fully comprehend how the trigger would behave.

File: sql/00000_base_database.sql (hunk 2 above)
Comment (Ian-Hodges) [reply sql/00000_base_database.sql]: Upped it to 1200.

File: sql/00003_test_result_audit_test_type_version_trigger.sql (hunk 3 above)
Comment (Ian-Hodges) [reply sql/00003_test_result_audit_test_type_version_trigger.sql:3]: Yes. Whether the deleted data saved to this table has any value, I couldn't say. Happy to remove it if redundant.

**Comment by Ian-Hodges** [comment]
> I have only one little concern around ever growing `testtype_version` table, as there is no retention on this table. Unless I may not fully comprehend how the trigger would behave.

I'll bring it up in the SU tomorrow. I would have thought the growth would be slow, but that is only a gut feeling.

**Review by ChrisPeska-BJSS — APPROVED** [review]

**Review by naathanbrown — APPROVED** [review]

**Pull request merged**